osu.Game/Rulesets/Edit/SelectionBlueprint.cs

```diff
@@ -129,6 +129,11 @@ protected virtual void OnSelected()
         /// </summary>
         public virtual Vector2 ScreenSpaceSelectionPoint => ScreenSpaceDrawQuad.Centre;
 
+        /// <summary>
+        /// The screen-space point that mark end of this <see cref="HitObjectSelectionBlueprint"/>.
+        /// </summary>
+        public virtual Vector2 ScreenSpaceEndPoint => ScreenSpaceDrawQuad.Centre;
```

Comment:
From a conceptual standpoint I am not sure this makes sense placed at this level. Not all selection blueprints will have defined "end points". A hitcircle selection blueprint has no real end - you could say that it "starts" and "ends" at the same point, but designing for this feels... too specific?

I think what may work better is to change `ScreenSpaceSelectionPoint` into a _collection_ of possible selection points. "Point objects" would return just the one (the centre of the draw quad). "Distance objects" could return many, and not even limited to start and end either - you could return selection points at each of the slider segments or something. Not sure we want to go there, but saying that at least it'd be _possible_, because that direction wouldn't hard-lock us into a start/end dichotomy as hard as what this branch is doing.

Doing this would even presumably get rid of some of the other awkward logic in this PR such as [this `m.ScreenSpaceSelectionPoint != m.ScreenSpaceEndPoint` check](https://github.com/ppy/osu/blob/788033e1b89997142f9d74089b19bd7821b08119/osu.Game/Screens/Edit/Compose/Components/BlueprintContainer.cs#L464).

@ppy/team-client would appreciate input here

Reply:
I agree this makes more sense, especially based on the logic required to make this work currently (the deduping done at the point of usage).

Reply:
I added `ScreenSpaceSelectionPoints`, many times we only care about the "main" selection point so I decided to leave `ScreenSpaceSelectionPoint` untouched. Also name `SelectionPoint`, at least for me, is a bit misleading. When I think about selection point I think about every point that select something. Maybe `MainSelectionPoint` or something like that?